traffic_ops/traffic_ops_golang/routes.go

```diff
@@ -162,15 +162,15 @@ func Routes(d ServerData) ([]Route, []RawRoute, http.Handler, error) {
 		//User: CRUD (1.4 then 1.1)
 		//Incrementing version for users because change to Nullable struct. Delete is completely new.
 		{1.4, http.MethodGet, `users/?(\.json)?$`, api.ReadHandler(user.GetTypeSingleton()), auth.PrivLevelReadOnly, Authenticated, nil},
-		{1.4, http.MethodGet, `users/{id}?(\.json)?$`, api.ReadHandler(user.GetTypeSingleton()), auth.PrivLevelReadOnly, Authenticated, nil},
-		{1.4, http.MethodPut, `users/{id}?(\.json)?$`, api.UpdateHandler(user.GetTypeSingleton()), auth.PrivLevelOperations, Authenticated, nil},
+		{1.4, http.MethodGet, `users/{id}$`, api.ReadHandler(user.GetTypeSingleton()), auth.PrivLevelReadOnly, Authenticated, nil},
+		{1.4, http.MethodPut, `users/{id}$`, api.UpdateHandler(user.GetTypeSingleton()), auth.PrivLevelOperations, Authenticated, nil},
 		{1.4, http.MethodPost, `users/?(\.json)?$`, api.CreateHandler(user.GetTypeSingleton()), auth.PrivLevelOperations, Authenticated, nil},
-		//{1.4, http.MethodDelete, `users/{id}?(\.json)?$`, api.DeleteHandler(user.GetTypeSingleton()), auth.PrivLevelOperations, Authenticated, nil},
+		//{1.4, http.MethodDelete, `users/{id}$`, api.DeleteHandler(user.GetTypeSingleton()), auth.PrivLevelOperations, Authenticated, nil},
 
 		{1.1, http.MethodGet, `users/?(\.json)?$`, handlerToFunc(proxyHandler), auth.PrivLevelReadOnly, Authenticated, []Middleware{}},
 		{1.1, http.MethodGet, `users/{id}?(\.json)?$`, handlerToFunc(proxyHandler), auth.PrivLevelReadOnly, Authenticated, []Middleware{}},
-		{1.1, http.MethodPut, `users/{id}?(\.json)?$`, handlerToFunc(proxyHandler), auth.PrivLevelReadOnly, Authenticated, []Middleware{}},
-		{1.1, http.MethodPost, `users/?(\.json)?$`, handlerToFunc(proxyHandler), auth.PrivLevelReadOnly, Authenticated, []Middleware{}},
+		{1.1, http.MethodPut, `users/{id}?(\.json)?$`, handlerToFunc(proxyHandler), auth.PrivLevelOperations, Authenticated, []Middleware{}},
```

Comment:
From routes.go

```
// 1.1 and 1.2 routes are simply a Go replacement for the equivalent Perl route. They may or may not conform with the API guidelines (https://cwiki.apache.org/confluence/display/TC/API+Guidelines).
// 1.3 routes exist only in a Go. There is NO equivalent Perl route. They should conform with the API guidelines (https://cwiki.apache.org/confluence/display/TC/API+Guidelines).
```

The idea here was that if you were simply rewriting a 1.1 Perl api endpoint in Go, it should be defined as 1.1 in routes.go to truly "override" it. It turns out that 99% of the routes in TrafficOpsRoutes.pm are 1.1 (only 1 is an 1.2 route - it's a weird asn one with a weird response structure).

So, imo, if i were to  look thru routes.go I should see:

- mostly 1.1 definitions because most are simply rewrites of Perl 1.1 endpoints
- one 1.2 definition that rewrites the "weird" asn endpoint
- some 1.3 endpoints that were new at the time
- some 1.4 endpoints that were new at the next time

my understanding is that every time something new is added (that is backwards compatible), we roll the minor version, 1.5, 1.6, etc...

but looking at routes.go, doesn't look like that was followed :(

Reply:
The minor version should only be incremented once per release though, right? E.g. if we release 3.0.0, master becomes "the next release". So if I add a new field to the cachegroups API, I would increment those minor versions. Somebody else comes along and adds yet another new field to the cachegroups API. At that point the version has already been incremented since the 3.0.0 release, so they wouldn't have to increment the version again?

Also, would the new incremented version have to follow the latest incremented version? E.g. cachegroups API is 1.1 right now. If I add a field, should it be incremented to 1.4 (since that's the "latest" version)?

To this day I am still confused about our API versioning and how things should really be handled and how the versioning corresponds to release versions. For instance, why do we need different handlers for v1.2 and v1.3 delivery services but not everything else? Is it because some backwards-incompatible changes were made, in which case it should really be v1.2 and v2.0 (except we hadn't truly followed SemVer yet)?

We do have _some_ info about versioning in the API guidelines on the wiki, but that doesn't really answer any of these questions to common examples. We should consider adding a `VERSIONING.md` to the traffic_ops_golang directory with just the API guidelines w.r.t. versioning requirements, explaining the common cases like adding a new optional field to an existing endpoint, rewriting a Perl endpoint, or implementing an entirely new endpoint. Also, any pre-SemVer exceptions should be documented (if any). Even the `README.md` in that directory seems to give contradictory information about the API versioning because it has an example of rewriting the `cdns` endpoint but uses 1.2 rather than 1.1.

I can take on the `VERSIONING.md` stuff if I can get some answers to these questions. We can't really expect people to follow the versioning requirements if they're not readily available and easy to understand.

Reply:
>The minor version should only be incremented once per release though, right?

Right.

>Also, would the new incremented version have to follow the latest incremented version? E.g. cachegroups API is 1.1 right now. If I add a field, should it be incremented to 1.4 (since that's the "latest" version)?

Right. The version is API-wide, not per-endpoint. An addition to `/cachegroups` needs to be at `/api/1.4/cachegroups`, **not** `/api/1.2/cachegroups`.

>We should consider adding a VERSIONING.md to the traffic_ops_golang directory

+1

We've had trouble remembering which is the current/next minor as well, and not adding new API minor versions when we should. I think maybe we should have a file somewhere, with which minor versions were added in which TC versions. Not sure the best way to ensure it's updated though; maybe a note in the Release Management wiki page?

Reply:
> my understanding is that every time something new is added (that is backwards compatible), we roll the minor  version, 1.5, 1.6, etc...
> 
> but looking at routes.go, doesn't look like that was followed :(
@mitchell852 

But if something in made in go, isn't it safer to increment the version since some changes can be made on accident and incrementing the version limits the number of things you need to look for? Breaking changes are easier to notice than minor changes.

If messages aren't the same across golang and perl would that count as a minor change? I'd have to look at it, but I'm thinking of validation messages in particular. I think the minor version should be incremented in cases of uncertainty at the very least.

Reply:
> Is the id portion still supposed to be optional?
> 
> Also, are we supposed to be still proxying 1.1 routes to the Perl even though those endpoints have now been rewritten in Go as 1.4? Should the Go routes be 1.1?  
@rawlinp 
It's not supposed to be optional. I'll update it.

I remember that the reverse proxy worked without an explicit route defined. However, by adding everything explicitly and removing the code that does the automatic reverse proxy we would be able to do something..

I'm forgetting what it was. It might have had something to do with implementing roles and capabilities or something related to priv_level. I thought a PR was made that has a list of all perl routes and if I'm not mistaken it was done by @rob05c?

Reply:
If there are Perl routes which need to be explicitly declared for some reason then that's probably fine to do in whatever PR for that new feature would be, but my understanding is that the endpoints rewritten in golang need to replace the Perl endpoints. If there are Perl routes explicitly declared as 1.1 and the golang rewrite at 1.4, then any clients requesting the 1.1 endpoint will still be hitting the Perl instead of the new golang version even though the golang is supposed to be backwards-compatible.

Reply:
> Any clients requesting the 1.1 endpoint will still be hitting the Perl instead of the new golang version even though the golang is supposed to be backwards-compatible.

Yes. Unless there is a hotfix that is needed for the current version, I think this is what is supposed to be done. To my knowledge we don't have any version number for patches. Incrementing a minor version should be done for backwards-compatible changes (see https://semver.org/).

Reply:
We kinda-sorta follow SemVer but not exactly (yet). Technically, nearly all of our golang rewritten APIs have made breaking changes by not including the same response fields as the Perl version when doing a POST/PUT, so really they should all be 2.x now until they've been truly fixed.

However, my concern is not really about the proper version number. My concern is that the Perl endpoints are still accessible even though the Golang endpoints are supposed to replace them completely. Old clients requesting 1.1 should be served by the rewritten golang version, since it's supposed to be backwards-compatible. We shouldn't serve Perl to 1.1 requests and Go to 1.2+ requests IMO. If there are no explicit routes to the Perl version, then I don't really care what the api version number is as long as it "makes sense".

Reply:
:+1: The proxy handler is - afaik - meant to handle cases where an endpoint hasn't been re-written, but the Go routes handler would shadow it if it were not explicitly declared. The idea is that if an endpoint has been re-written in a way that's consistent with the behavior of a 1.1 endpoint, that 1.1 endpoint should no longer be served by Perl code.

Reply:
Hmm.. interesting.

Reply:
>The proxy handler is - afaik - meant to handle cases where an endpoint hasn't been re-written, but the Go routes handler would shadow it if it were not explicitly declared.

Right now, but FYI that will be changing once Capabilities are enforced (https://github.com/apache/trafficcontrol/pull/2791). After that, no Perl route without a Go declaration will be proxied. Just FYI.